sqld/src/namespace.rs

```diff
@@ -0,0 +1,304 @@
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_lock::{RwLock, RwLockUpgradableReadGuard};
+use bytes::Bytes;
+use hyper::Uri;
+use tokio::sync::mpsc;
+use tokio::task::JoinSet;
+use tonic::transport::Channel;
+
+use crate::database::config::DatabaseConfigStore;
+use crate::database::dump::loader::DumpLoader;
+use crate::database::factory::{DbFactory, TrackedDb};
+use crate::database::libsql::{LibSqlDb, LibSqlDbFactory};
+use crate::database::write_proxy::{WriteProxyDatabase, WriteProxyDbFactory};
+use crate::database::Database;
+use crate::replication::primary::logger::{ReplicationLoggerHookCtx, REPLICATION_METHODS};
+use crate::replication::replica::Replicator;
+use crate::replication::{NamespacedSnapshotCallback, ReplicationLogger};
+use crate::stats::Stats;
+use crate::{
+    check_fresh_db, init_bottomless_replicator, run_periodic_compactions, DB_CREATE_TIMEOUT,
+    MAX_CONCURRENT_DBS,
+};
+
+#[async_trait::async_trait]
+pub trait NamespaceFactory: Sync + Send + 'static {
```

Comment:
I think this trait name is super confusing. I would maybe call it `NamedspacedDb` and just drop the factory since I think we all would know this is a factory :D (or can call it Maker which feels more rusty).